Update ROOT's llvm to llvm13

smuzaffar opened the pull request.

vgvassilev committed:
Add getSizeExpr parameter.

See llvm-project/llvm@772e266fbff40a566c9ef2d805c057c4692ac028

vgvassilev committed:
llvm::StringRef does not support automatic conversion to string anymore.

See llvm/llvm-project@777180a32b61

vgvassilev committed:
Color handling moved from raw_fd_ostream to raw_ostream.

We do not need the wrapper which did more than needed.

See llvm/llvm-project@8744d7f25b64000f8e34bc8e2c8b8eaccb4c7910

vgvassilev committed:
The mangler accepts GlobalDecls and we don't need special cases ctors and dtors.

See llvm/llvm-project/29e1a16be8216066d1ed733a763a749aed13ff47

vgvassilev committed:
Preprocessor::LookupFile returns FileEntryRef.

See llvm/llvm-project@4dc5573acc0d2e7c59d8bac2543eb25cb4b32984

vgvassilev committed:
getName returns now Expected<StringRef>

See llvm/llvm-project@bcc00e1afbdc107e05b2977fde26053a52295dd1

vgvassilev committed:
CPlusPlus2a was renamed to CPlusPlus20.

vgvassilev committed:
llvm::sys::fs went into a separate header.

vgvassilev committed:
CreateFromArgs takes now an ArrayRef.

See llvm/llvm-project@1fac68b0dc19b03fd2c5e9856f0f2c5a11691348

vgvassilev committed:
ShowTimers went in CodeGenOptions.

See llvm/llvm-project@1821265db681cd2289fce9331e3aed26bdf814e3

vgvassilev committed:
OmitFramePointer flag became an enum.

See llvm/llvm-project@ff22ec3d7004ada2ef11893c832624fdd0caa5c6

vgvassilev committed:
The BuiltinContext is in a separate header.

vgvassilev committed:
getFlags returns llvm::Expected.

vgvassilev committed:
FileManager::getFile returns an llvm::ErrorOr which needs to be checked.

vgvassilev committed:
Qualify DestroyTemplateIdAnnotationsRAIIObj.

vgvassilev committed:
ContentCache::getRawBuffer is replaced with getBufferDataIfLoaded

See llvm/llvm-project@156e8b37024abd8630666e0ae8a251b80299ed1d

vgvassilev committed:
FileInfo always has content cache and replaceBuffer became setBuffer.

See llvm/llvm-project@296314516d103f8eeb987a08b509c1381cfeef89

vgvassilev committed:
createOutputFile was simplified.

See llvm/llvm-project@ad7aaa475e5e632242b07380ec47d2f35d077209

vgvassilev committed:
stripPointerCastsNoFollowAliases got renamed to stripPointerCasts

See llvm/llvm-project@2452d7030b84d358165f5f76c9270937b4ea1a91

vgvassilev committed:
Expr dependence was reworked and made protected.

By default we should get no dependence, assert that instead.

See llvm/llvm-project@ec3060c72de6ab6992269318d92764199856e5fe

vgvassilev committed:
dump() takes an ASTContext now.

See llvm/llvm-project@473fbc90d1fbf17e16594dfb17248c4f50b4d399

weliveindetail committed:
Implement --debug-only option in order to easily dump LLVM debug output

For example: `cling --debug-only="orc,jitlink"`

vgvassilev committed:
CGFT_AssemblyFile was moved to the llvm namespace.

See llvm/llvm-project@1dfede3122eec83b885d788553e3620806d54650

vgvassilev committed:
The StoredDeclsList is reworked and more user-friendly.

See llvm/llvm-project@0cb7e7ca0c864e052bf49978f3bcd667c9e16930

vgvassilev committed:
createDriverOptTable became a singleton.

See llvm/llvm-project@433927595dde3821c76b33725aab8ccd472a5137

vgvassilev committed:
llvm::make_unique -> std::make_unique

vgvassilev committed:
Add an include to TargetInfo.h

vgvassilev committed:
Use the real name not it's synonym.

See llvm/llvm-project@518d955f9dd2f5f854006a71ff9c8b117a66548b

weliveindetail committed:
Rework the cling JIT to use the ORCv2 interface.

The ORCv1 JIT was deprecated and removed in llvm13. The ORCv2 provides several
benefits such as parallel compilation.

vgvassilev committed:
Dropped compatibility flags F_{Text,None} -> OF_{Text,None}

See llvm/llvm-project@3302af9

vgvassilev committed:
TemplateArgument::print takes one more argument.

See llvm/llvm-project@99d63ccff04b672694f8a2b3eed024b873dc163d

vgvassilev committed:
Add the override qualifier to suppress warnings

vgvassilev committed:
VK_RValue -> VK_PRValue

See llvm/llvm-project@aef5d8fdc7d0d348125d5ecf4a13be5888eb1654

vgvassilev committed:
createModuleManager and friends was renamed to *ASTReader.

See llvm/llvm-project@20d51b2f14ac4488f684f8fc57cb0ba718a6b91d

vgvassilev committed:
Use the synonym interface getBuffer->getBufferData

vgvassilev committed:
TypeNameContext was renamed to TypeName.

See llvm/llvm-project@e4d27932a59fb61aaba3ff7a3ccd1b5bc9215fb9

vgvassilev committed:
getTypeAnnotation returns now TypeResult.

See llvm/llvm-project@330873230071ffc2aebc0fe74db55e7a530c2f1b

vgvassilev committed:
APSInt::toString returning std::string was removed.

See llvm/llvm-project@61cdaf66fe22be2b5942ddee4f46a998b4f3ee29

vgvassilev committed:
setSuppressAllDiagnostics has no default anymore.

Remove duplicate line of code.

vgvassilev committed:
ParseUnqualifiedId has new set of params.

See llvm/llvm-project@0dd0b1017c807e7a803e69be7bdc5c0ac7b469d8

vgvassilev committed:
DisablePCHValidation became an enum

See llvm/llvm-project@b0e89906f5b7e

vgvassilev committed:
DirectoryLookup takes a DirectoryRef.

See llvm/llvm-project@0377ca641c9af0262270d091578cab19523b1bc6

vgvassilev committed:
CreateASTDumper takes another bool arg.

See llvm/llvm-project@e135cf8a03b97

vgvassilev committed:
Target::adjust takes DiagnosticsEngine

vgvassilev committed:
LookupFile returns FileEntryRef.

vgvassilev committed:
OPT_fmodule_name was dropped.

vgvassilev committed:
Use AnnotateAttr::CreateImplicit function which is now available.

See llvm/llvm-project@d093401a2617d3c46aaed9eeaecf877e3ae1a9f1

vgvassilev committed:
PrintHelp is now lower case.

vgvassilev committed:
llvm::StringRef::*lower became insensitive.

See llvm/llvm-project@2e4a2b8430aca6f7aef8100a5ff81ca0328d03f9

vgvassilev committed:
The new API requires the address of the files

vgvassilev committed:
Connect the JIT with the lazy symbol creation infrastructure.

vgvassilev committed:
OrcV2 is more strict with linkage.

When we compile incremental inputs we store each input in a separate
llvm::Module. In order to preserve the semantics, we make all symbols from
each module with external linkage, that is, accessible from other modules.

This model works well for almost all symbols, except for the clang-generated
ones which have the same name. For example clang stores strings in variables
`.strN` which may clash accross modules. Luckily, these strings are used within
a single module and thus we can avoid the JIT error by keeping them with their
non-external linkage.

vgvassilev committed:
Drop the unused Interpreter::addSymbol function.

It just adds to the complexity and confusion of the addSymbol implementation

vgvassilev committed:
Implement a solution to the JIT-owned llvm::Modules.

When removing code, we need to recover Clang's CodeGen and this happens by
iterating over the clang::Decls and the llvm::Module for a given Transaction in
the TransactionUnloader.

The problem is that now OrcV2 takes the ownership of the llvm::Module from the
Transaction for itself and then it can notify us when the module was compiled on
demand. Unfortuantely, there is no good way to guarantee that the same module
will be attached to the same transaction. This approach allows the
TransactionUnloader::unloadModule logic to recover CodeGen.

In a longer term, when we switch to partial translation units from clang-repl
we might be able to recover CodeGen without needing to know details about the
generated llvm::Module.

vgvassilev committed:
IncrementalJIT::getSymbolAddress is expected to return 0 and never complain.

vgvassilev committed:
Jitlink does not support all necessary platfroms, switch to RuntimeDyld.

vgvassilev committed:
Rework the symbol injection mechanism to also allow symbol replacement.

vgvassilev committed:
The OrcV2 now can iterate over the globals and run the static init for us.

vgvassilev committed:
Update the comment, we still need this if g++ > 5

vgvassilev committed:
Print a<b<c> >' rather than 'a<b<c>>' in the Transform.C test.

See llvm/llvm-project@159a9f7e76307734bcdcae3357640e42e0733194

vgvassilev committed:
FIXME: Undo this change and debug why we have PendingInstances.

vgvassilev committed:
LLVM now requires C++14.

Patch by Lang Hames.

vgvassilev committed:
Rename addSymbol to match what it does and remove redundant params.

vgvassilev committed:
Diagnostic text changed

vgvassilev committed:
Re-teach getAddressOfGlobal to tell if something came from the JIT.

This patch also inverts the parameter from ExcludeHostSymbols to
IncludeHostSymbols to improve readability.

As written in one of the FIXMEs this is a temporary solution which allows us to
go further with the llvm-13 upgrade. We can rework the getAddress* functionality
to reuse more the ORCV2 infrastructure and more importantly to use that
infrastructure in the recommended way.

vgvassilev committed:
Unconditionally install ClingTargets.cmake, required by ClingConfig.cmake

This fixes our -Dbuiltin_cling=Off option.

vgvassilev committed:
Complete the upgrade to LLVM/Clang 13 for cling.

vgvassilev committed:
APInt/APSInt toString() std::string variants were removed for llvm::toString

See llvm/llvm-project@61cdaf66fe22be2b5942ddee4f46a998b4f3ee29

vgvassilev committed:
isIntegerConstantExpression was refactored to getIntegerConstantExpression.

See llvm/llvm-prject@36036aa70ec1df7b51b5d30b2dd8090ad2b6e783

vgvassilev committed:
Improve the builtin_cling=Off feature.

vgvassilev committed:
ASTContext::getConstantArrayType has a new param called SizeExpr.

See llvm/llvm-project@772e266fbff40a566c9ef2d805c057c4692ac028

vgvassilev committed:
IsMissingRequirement was renamed to IsUnimportable.

See llvm/llvm-project@fc76b4ad3da3c1f5fdc34370a55d613b3ca7af73

vgvassilev committed:
getClangToolFullVersion doesn't show empty space anymore.

See llvm/llvm-project@bbea4d5e6b82a683dccaa8f4916e2a44f5dd3490

vgvassilev committed:
raw_string_ostream:: takes an ASTContext.

vgvassilev committed:
[llvm13] -fmodule-name requires now `=`.

vgvassilev committed:
Spell out correctly the forward decl

vgvassilev committed:
getContentCache() returns a reference not a pointer.

vgvassilev committed:
FileManager::getDirectory returns llvm::Error, use getOptionalDirectoryRef

vgvassilev committed:
Use consistent TUKind for the Preproocessor and Sema.

This fixes an assertion.

vgvassilev committed:
Remove redundant header include.

vgvassilev committed:
Move handling of the help option alias after hinding all irrelevant options.

This fixes an option double registration assert.

vgvassilev committed:
Add another symbol generator to resolve the generated lazy symbol.

The in-process symbol generator resolves symbols available in the process and
all dlopened libraries. However, upon unresolved symbol, we trigger our lazy
loading function creator callbacks which load the relevant library. Then we
need another in-process symbol generator to catch the newly created symbol so
that the JIT is happy.

This solution is suboptimal and in the comments is provided a cleaner
implementation, however, this implementation loads libraries for weak symbols
that the JIT could generate. This needs to be further investigated.

vgvassilev committed:
Complete the llvm13 upgrade for ROOT.

vgvassilev committed:
Import llvm13.

vgvassilev committed:
[cmake] Force-set the ZLIB_LIBRARY

Apparently cmake sets the cache after the configuration is complete and using
FORCE updates the cache at the spot including the relevant generator expressions.

This fixes the case where we build ROOT with -Dbuiltin_zlib=Off and then LLVM
sanity checks the found zlib issuing:

```
Error evaluating generator expression:

  $<TARGET_FILE:ZLIB>

No target "ZLIB"
```

vgvassilev committed:
Do not install {Clang,Cling}Config.cmake in the project lib dir.

This should fix the clad builds and unpollute the ROOTSYS/lib.

vgvassilev committed:
Try fixing the multicore

vgvassilev committed:
Use the header fullpath to fix a new diagnostic.

When creating a dictionary we should either copy the header next to the
modulemap file (in the build folder) or use full paths.

This patch fixes new diagnostics from such kind:

tree/dataframe/test/module.modulemap:3:33: error: cannot emit module dataframe_splitcoll_arrayview: size must be explicitly specified for missing header file "TwoFloats.h"
  module "TwoFloats.h" { header "TwoFloats.h" export * }
                                ^

hahnjo committed:
Fix build with C++17

When building with C++17, a llvm::StringRef can be implicitly converted
to std::string_view. This causes an ambiguity when comparing a StringRef
with a const char * because the compiler could either
 * convert the StringRef into a std::string_view and use the operator==
   from TString.h, taking std::string_view and const char*, or
 * construct a StringRef from the const char* and use the opertor==
   defined in StringRef.h.

In TCling.cxx, we want to compare two StringRefs and can construct the
StringRefs statically beforehand.

hahnjo committed:
Fix CUDA architecture selection

--cuda-gpu-arch aliases --offload-arch, so the option parser will
canonicalize to the latter.

hahnjo committed:
[CUDA/HIP] Remove argument from module ctor/dtor signatures

This works around a bug + missing support in ORC for constructors with
non-void signatures (https://github.com/llvm/llvm-project/issues/54797).
Also submitted upstream, see https://reviews.llvm.org/D123370

hahnjo committed:
[cling] Do not externalize unnamed GlobalValues

It doesn't make sense and the JIT crashes, see upstream bug report
https://github.com/llvm/llvm-project/issues/54813

hahnjo committed:
[cling] Fix some trivial compiler warnings

"moving a local object in a return statement prevents copy elision" as
well as an unused variable and function referenced only from an assert.

hahnjo committed:
[cling] Move Autoload annotation after scope keyword

Otherwise tests/Autoloading/AutoForwarding.C complains that "an
attribute list cannot appear here".

hahnjo committed:
[cling] Temporarily disable ReuseExistingWeakSymbols

This pass calls getSymbolAddress which is a lot more expensive with
the upgrade to LLVM 13. Disable for now to speed up test execution
and avoid a number of timeouts. Investigate later if it's possible
to re-enable.

hahnjo committed:
[cling] Print 'a<b<c> >' rather than 'a<b<c>>' in ValuePrinter

See earlier commit for the Transform.C test.

vgvassilev committed:
[cling] The JIT now is more consistent on OSX with the linker-level mangling

OSX (due to legacy to disambigate between asm and C symbols) implements an extra
linker-level mangling which adds another `_` in the name. However, that is only
in the on-disk representation and needs to be dropped upon calling dlsym.

OrcV2 gives us a handle to the symbol with starts with `__` allowing us to
invert the logic of the symbol search where we drop the `__` for dlsym.

vgvassilev committed:
Fix the special case for Windows. FIXE: Try remove ifdef

hahnjo committed:
[cling] Fix emission of vtables

It is (again) necessary to call DefineUsedVTables() to enable (at
least) the emission of implicitly defined destructors overriding
the virtual destructor in a base class, see the added test.

vgvassilev committed:
Print 'a<b<c> >' rather than 'a<b<c>>' in TCling.

See earlier commit for the cling/test/.../Transform.C.

hahnjo committed:
[cling] Ignore duplicate kept GlobalValues

Set WeakAnyLinkage instead of ExternalLinkage to tell the JIT linker
that it should not error on duplicate symbols.

hahnjo committed:
TClingMethodInfoTests: Delete unwanted constructor

Starting with version 12, Clang seems to implicitly generate the
Derived() constructor, calling Base1() and then Base2(). Older
versions and GCC complain that the "call of overloaded 'Derived()'
is ambiguous". It's not clear to me what is actually the correct
behavior according to the C++ standard, so just delete the unwanted
constructor for now.

hahnjo committed:
[cling] Print default template arguments

Fixes (at least) roottest-cling-stl-default-VectorSort and
roottest-cling-templateMembers-runoperatorEqual.

hahnjo committed:
[cling] Avoid duplicate symbols from RuntimeUniverse.h

Make InterpreterGeneratedCodeDiagnosticsMaybeIncorrect static, which
is then turned into a weak symbol by KeepLocalGVPass.

hahnjo committed:
TCling: Print default template arguments

Fixes the build of generated dictionaries.

vgvassilev committed:
[D126620][clang] AST/Print: honor AlwaysIncludeTypeForTemplateArgument policy.

This redoes D103040 in a way that AlwaysIncludeTypeForTemplateArgument = false
policy is honored for printing template specialization types.
This can be seen for example when printing a canonicalized
dependent TemplateSpecializationType which has integral arguments.

Signed-off-by: Matheus Izvekov <mizvekov@gmail.com>

This fixes:
before:
root.exe -l -b -q -e 'TClass::GetClass("std::array<int,3>")->Print()'
OBJ: TClassarray<int,3UL>

after: root.exe -l -b -q -e 'TClass::GetClass("std::array<int,3>")->Print()'
OBJ: TClassarray<int,3>

hahnjo committed:
[cling] Weaken definitions of typeinfos and vtables

Introduce a backend pass that changes the linkage of external symbols
starting with "_ZT" (ie. typeinfo names, typeinfos, and vtables) to
weak symbols. This avoids duplicate symbol errors with the new JIT
linker.

FIXME: This is a hack, we should teach the frontend to emit these
only once, or mark all duplicates as available_externally (if that
improves performance due to optimizations).

vgvassilev committed:
LLVM has now unnamed for enums and anonymous for everything else.

See llvm-project/llvm@50542d504dd869fe1241a9cc987b72ead5a56073

vgvassilev committed:
Revert "Reland another version of "The Sema::LookupConstructor is not iteration safe.""

This reverts commit 6494f1f34e7eb6a5cc71c7ae808d4728bba13e3e.

LLVM13 has fixed the issue of deserialization while performing lookups.

See llvm/llvm-project@0cb7e7ca0c864e052bf49978f3bcd667c9e16930

saisoma123 committed:
Added cmakeDir for Windows

cmsbuild commented:
A new Pull Request was created by  @smuzaffar (Malik Shahzad Muzaffar) for branch cms/master/6f834eb.

@cmsbuild, @smuzaffar, @aandvalenzuela, @iarspider can you please review it and eventually sign? Thanks.
@perrotta, @dpiparo, @qliphy you are the release manager for this.
cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

hahnjo committed:
[cmake] Reset variable in ROOT_FIND_DIRS_WITH_HEADERS (#10682)

This makes sure that the list of directories is actually empty in
case neither inc/ nor v7/inc/ exist, and does not contain left-overs
that were stored in this variable in a parent scope.

After this change, roottest/root/meta/rootcling/module.modulemap is
much shorter and in particular free of non-existent entries complained
about by the LLVM 13 upgrade.

hahnjo committed:
Revert "Add support for profiling/debugging interpreted/JITted code"

This reverts commit 22b1606f5fd842590f04724d789de5b29495efc6 until it
is re-implemented on top of the changes for upgrading to LLVM 13.

cmsbuild commented:
Pull request #170 was updated.

smuzaffar commented:
test parameters:
- full_cmssw = true

smuzaffar commented:
please test for CMSSW_12_5_ROOT6_X

cmsbuild commented:
-1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6342e6/25491/summary.html
**COMMIT**: 8e20719c827b49f991254816891d2c5e65d74bbe
**CMSSW**: CMSSW_12_5_ROOT6_X_2022-06-12-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/root/170/25491/install.sh` to create a dev area with all the needed externals and cmssw changes.


## External Build

I found compilation error when building: 

<pre>Requested to quit.
* The action "build-external+py3-hepdata-lib+0.8.1-ec986c2493f655785ae2df5c3741f2df" was not completed successfully because Failed to build py3-hepdata-lib. Log file in /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc10/external/py3-hepdata-lib/0.8.1-ec986c2493f655785ae2df5c3741f2df/log. Final lines of the log file:
'"'"''"'"''"'"' % ('"'"'/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc10/external/py3-hepdata-lib/0.8.1-ec986c2493f655785ae2df5c3741f2df/cmsdist-tmp/pip-req-build-02hlcwf_/setup.py'"'"',), "<pip-setuptools-caller>", "exec"))' egg_info --egg-base /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc10/external/py3-hepdata-lib/0.8.1-ec986c2493f655785ae2df5c3741f2df/cmsdist-tmp/pip-pip-egg-info-pxqkc5sy
cwd: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc10/external/py3-hepdata-lib/0.8.1-ec986c2493f655785ae2df5c3741f2df/cmsdist-tmp/pip-req-build-02hlcwf_/
Preparing metadata (setup.py): finished with status 'error'
error: metadata-generation-failed

 Encountered error while generating package metadata.
> See above for output.

note: This is an issue with the package mentioned above, not pip.
</pre>

smuzaffar commented:
please test with https://github.com/cms-sw/cmsdist/pull/7931 for CMSSW_12_5_ROOT6_X

cmsbuild commented:
-1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6342e6/25499/summary.html
**COMMIT**: 8e20719c827b49f991254816891d2c5e65d74bbe
**CMSSW**: CMSSW_12_5_ROOT6_X_2022-06-12-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/root/170/25499/install.sh` to create a dev area with all the needed externals and cmssw changes.


## External Build

I found compilation error when building: 

<pre>Requested to quit.
* The action "build-external+py3-hepdata-lib+0.8.1-23017aefcaec1f6e033b49c0652b0cd2" was not completed successfully because Failed to build py3-hepdata-lib. Log file in /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc10/external/py3-hepdata-lib/0.8.1-23017aefcaec1f6e033b49c0652b0cd2/log. Final lines of the log file:
'"'"''"'"''"'"' % ('"'"'/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc10/external/py3-hepdata-lib/0.8.1-23017aefcaec1f6e033b49c0652b0cd2/cmsdist-tmp/pip-req-build-n9zun6iw/setup.py'"'"',), "<pip-setuptools-caller>", "exec"))' egg_info --egg-base /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc10/external/py3-hepdata-lib/0.8.1-23017aefcaec1f6e033b49c0652b0cd2/cmsdist-tmp/pip-pip-egg-info-oa88x8hh
cwd: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc10/external/py3-hepdata-lib/0.8.1-23017aefcaec1f6e033b49c0652b0cd2/cmsdist-tmp/pip-req-build-n9zun6iw/
Preparing metadata (setup.py): finished with status 'error'
error: metadata-generation-failed

 Encountered error while generating package metadata.
> See above for output.

note: This is an issue with the package mentioned above, not pip.
</pre>
